src/libraries/api/client/dto/proposal.js

```
@@ -0,0 +1,24 @@
+// @flow
+import ServiceDefinitionDto from './service-definition'
+
+class ProposalDto {
+  id: number
+  providerId: string
+  serviceType: string
+  serviceDefinition: {
+    locationOriginate: {
+      country: ?string
+    }
+  }
+
+  constructor (data: Object) {
+    this.id = data.id
+    this.providerId = data.providerId
+    this.serviceType = data.serviceType
+    if (data.serviceDefinition) {
```

Not sure about this, since a proposal cannot be without a definition. Maybe throw an error if it's unset?

---

If this DTO can contain `undefined` values, when we have to be aware about that everywhere we're using it - which is bad.
I agree that constructor should not throw errors.

What about having an `isValid(): boolean` method in `ProposalDTO`, which would return `true` if all required values are in place? We could check this value right after constructing dto:

```js
const dto = new ProposalDTO(...)
if (!dto.isValid()) {
  throw new Error("Missing required proposal fields")
}
```

---

Problem is not related to `if (data.serviceDefinition)`, undefined can happen for all fields e.g. `this.id = data.id = undefined`